R/TealAppDriver.R

```diff
@@ -0,0 +1,427 @@
+          sprintf(
+            "#%s-active-filter_active_vars_contents > span",
+            self$get_active_ns("filter_panel")
+          )
+        ),
+        function(x) {
+          style <- x %>%
+            rvest::read_html() %>%
+            rvest::html_node("span") %>%
+            rvest::html_attr("style")
+          style <- ifelse(is.na(style), "", style)
+          style != "display: none;"
+        }
+      ) %>%
+        unname()
+      available_datasets <- self$app$get_html(
+        sprintf(
+          "#%s-active-filter_active_vars_contents",
+          self$get_active_ns("filter_panel")
+        )
+      ) %>%
+        read_html() %>%
+        html_nodes(".filter_panel_dataname") %>%
+        html_text()
+      available_datasets[displayed_datasets_index]
+    },
+    #' @description
+    #' Get the active filter variables from a dataset in the `teal` app.
+    #'
+    #' @param dataset_name (character) The name of the dataset to get the filter variables from.
+    get_active_data_filters = function(dataset_name) {
+      sapply(
+        self$app$get_html(
+          sprintf(
+            "#%s-active-%s-filter-cards .filter-card-varname",
+            self$get_active_ns("filter_panel"),
+            dataset_name
+          )
+        ),
+        function(x) {
+          x %>%
+            rvest::read_html() %>%
+            rvest::html_text() %>%
+            gsub(pattern = "\\s", replacement = "")
+        }
+      ) %>%
+        unname()
+    },
+    #' @description
+    #' Get the active filter values from the active filter selection of dataset from the filter panel.
+    #'
+    #' @param dataset_name (character) The name of the dataset to get the filter values from.
+    #' @param var_name (character) The name of the variable to get the filter values from.
+    #' @param is_numeric (logical) If the variable is numeric or not.
+    #'
+    #' @return The value of the active filter selection.
+    get_filter_selection_value = function(dataset_name, var_name, is_numeric = FALSE) {
+      selection_suffix <- ifelse(is_numeric, "selection_manual", "selection")
+      self$app$get_value(
+        input = sprintf(
+          "%s-active-%s-filter-%s_%s-inputs-%s",
+          self$get_active_ns("filter_panel"),
+          dataset_name,
+          dataset_name,
+          var_name,
+          selection_suffix
+        )
+      )
+    },
+    #' @description
+    #' Add a new variable from the dataset to be filtered.
+    #'
+    #' @param dataset_name (character) The name of the dataset to add the filter variable to.
+    #' @param var_name (character) The name of the variable to add to the filter panel.
+    #'
+    #' @return The `TealAppDriver` object invisibly.
+    add_filter_var = function(dataset_name, var_name) {
+      self$app$set_inputs(
+        !!sprintf(
+          "%s-add-%s-filter-var_to_add",
+          self$get_active_ns("filter_panel"),
+          dataset_name
+        ) := var_name
+      )
```

Comment:
doesn't do.call require to put parameters in a list, which does not resolve the issue of the name of the parameter being a pasted string?

Reply:
@m7pr yes:
```r
do.call(
  self$app$set_inputs, 
  setNames(
    list(varname), 
    sprintf(
       "%s-add-%s-filter-var_to_add",
       self$get_active_ns("filter_panel"),
       dataset_name
      )
   )
)
```

Reply:
Do you mean providing the ability to add multiple vars together?

I'm not fully convinced if `do.call` is the way to do this. Primarily because this encourages people to store their vars in a variable and use it and this can be hard to debug when the test will fail. So, I would like an implementation that involves allowing people to explicitly set the inputs that they wish to change [like how it's done in `AppDriver$set_inputs()`](https://rstudio.github.io/shinytest2/reference/AppDriver.html#method-set-inputs-)

Reply:
Looks like I was thinking of something else entirely. Thanks for the code example. This looks good.

Reply:
Added a public [`set_input`](https://github.com/insightsengineering/teal/pull/1127/commits/2d13df0b80320f21551c964094c072e2867f8004) method to help set inputs based on the suggestion. Will replace all the instances of `AppDriver$set_inputs` with this wrapper.

Reply:
thanks, this removes the `!!` and `:=`. `setNames` now looks so obvious : P

Reply:
@vedhav @m7pr it is more about operators which I don't like. We avoided any non-base operators whenever it is possible on production (except magrittr `%>%` which is used in reproducible modules code).

However, it is 2024 and I think I should move on ;)

```r
library(rlang)
foo <- function(...) return(list2(...))
arg_name <- "a"

microbenchmark::microbenchmark(
  rlang_bang2 = foo(!!arg_name := TRUE),
  do.call = do.call(foo, setNames(list(TRUE), arg_name)),
  times = 10000L
)
# Unit: nanoseconds
#         expr  min   lq     mean median   uq   max neval cld
#  rlang_bang2  902  984 1106.857    984 1025 38991 10000  a 
#      do.call 1599 1681 1953.859   1722 1886 27265 10000   b
```

👍